Issue 9: Getting started vignette (#32)
epinowcast/epidist: vignettes/epidist.Rmd

  error = FALSE
)
```

```{r load-requirements}
library(epidist)
library(data.table)
library(purrr)
library(ggplot2)
```

Many quantities in epidemiology can be thought of as the time between two events, or "delays".
Important examples include:

* the incubation period (time between infection and symptom onset),
* serial interval (time between symptom onset of infectee and symptom onset of infected), and 
* generation interval (time between infection of infectee and infection of infected).

Unfortunately, estimating delays accurately from observational data is usually difficult.
In our experience, the two main issues are:

1. interval censoring, and
2. right truncation.

Don't worry if you've not come across these terms before!
In Section \@ref(data), we will explain what they mean by simulating example data analogous to what we might observe during an ongoing infectious disease outbreak.
Next, in Section \@ref(fit), we show how `epidist` can be used to estimate delays using a statistical model which properly accounts for these two issues.
Finally, in Section \@ref(compare), we demonstrate that the fitted delay distribution accurately recovers the underlying truth.

> 

Reviewer: ```suggestion

```suggestion

If you want more technical details: the `epidist` package implements models following best practices described in [Park et al. 2023](https://www.medrxiv.org/content/10.1101/2024.01.12.24301247v1) and [Charniga et al. 2024](https://arxiv.org/abs/2405.08841).
```
```

Author: Agree. Added with Bibtex. Arguably you might like the link?

Reviewer 2: I think either or works. Personally I find links more useful but I get the portability arg behind just giving a citation

# Example data {#data}

> Data should be in a certain format for use within the `epidist` package.

Reviewer: Here, I'd insert a few lines describing the required data schema, including the required fields, a definition for each field, and notes on data type requirements, e.g.:

```suggestion
Data should be formatted as a [`data.table`]() with the following columns for use within the `epidist` package:

* `case` (data type): a unique ID for each observed case
* `ptime` (data type): ...
* etc. 
```

Author: I think this is a good suggestion. My only uncertainty is that I don't exactly know which columns are "hard requirements" for using the package and which are "there sometimes".

I suppose doing this will clarify this point for me (which is needed to do anyway).

Author: I still think this is a good suggestion but would like to leave to a later issue as I am unsure of the input requirements we want, and also unsure of the columns we want.

```{r outbreak, fig.cap="(ref:outbreak)"}
outbreak[case %% 50 == 0, ] |>
  ggplot(aes(x = ptime, y = case)) +
  geom_point(col = "#56B4E9") +
  labs(x = "Primary event time (day)", y = "Case number") +
  theme_minimal()
```

`outbreak` is a `data.table` with columns for the case number `case` and the time of primary event `ptime`.

Now, to generate secondary events, we will use a lognormal distribution (Figure \@ref(fig:lognormal)) for the delay between primary and secondary events:

```{r}
secondary_dist <- data.table(meanlog = 1.8, sdlog = 0.5) |>
  add_natural_scale_mean_sd()
```

(ref:lognormal) The lognormal distribution is skewed to the right. Long delay times still have some probability.

```{r lognormal, fig.cap="(ref:lognormal)"}
ggplot(data.frame(x = c(0, 8)), aes(x = x)) +
  geom_function(
    fun = dlnorm,
    meanlog = secondary_dist[["meanlog"]],
    sdlog = secondary_dist[["sdlog"]]
  ) +
  theme_minimal() +
  labs(
    x = "Delay between primary and secondary event (days)",
    y = "Probability density"
  )
```

```{r}
obs <- outbreak |>
  simulate_secondary(
    meanlog = secondary_dist[["meanlog"]],
    sdlog = secondary_dist[["sdlog"]]
  )
```

> (ref:delay) Secondary events (in green) occur at random  (As with Figure \@ref(fig:outbreak), to make this figure easier to read, only every 50th case is shown.)

Reviewer: I think many non-statistical users see "at random" and think "unpredictably", but you're really trying to convey that this is a stochastic process. Consider rewording?

Author: "occur with a delay drawn from the lognormal distribution" --> chance to reinforce the delay distribution

Reviewer 2: Good point. I like the reword suggestions

```{r delay, fig.cap="(ref:delay)"}
obs[case %% 50 == 0, ] |>
  ggplot(aes(y = case)) +
  geom_segment(
    aes(x = ptime, xend = stime, y = case, yend = case),
    col = "grey"
  ) +
  geom_point(aes(x = ptime), col = "#56B4E9") +
  geom_point(aes(x = stime), col = "#009E73") +
  labs(x = "Event time (day)", y = "Case number") +
  theme_minimal()
```

`obs` is a `data.table` object with additional columns for the delay `delay` and time of secondary event `stime`.
The secondary event time is simply the primary event time plus the delay:

```{r}
all(obs$ptime + obs$delay == obs$stime)
```

If we received the data `obs` as above then estimating the delay distribution would be easy, and this package wouldn't need to exist.
However, in reality, we will almost never receive the data as above.

First, the times of primary and secondary events will usually be censored.
This means that rather than exact event times, we observe event times within an interval.

> Here we suppose that the interval is daily, resulting in an obscuring of the exact delay data (Figure \@ref(fig:cens)):

Reviewer: ```suggestion
Here we suppose that the interval is daily, meaning that only the date of the primary or secondary event, not the exact time of each event, is reported.  (Figure \@ref(fig:cens)):
```

Author: Agree.

```{r}
# observe_process() should be renamed and include a "daily" i.e. 1 argument
obs_cens <- obs |> observe_process()
```

(ref:cens) Interval censoring of the primary and secondary event times obscures the delay times.

```{r cens, fig.cap="(ref:cens)"}
ggplot(obs_cens, aes(x = delay, y = delay_daily)) +
  geom_point(col = "#E69F00") +
  geom_abline(slope = 1, intercept = 0, linetype = "dashed", col = "grey") +
  theme_minimal() +
  coord_fixed() +
  labs(x = "Exact delay time (days)", y = "Censored delay time (days)")
```

> Next, during an outbreak we will usually be estimating delays in real time.

Reviewer: Epidemiological delays are often ascertained via the secondary event (e.g. when estimating an incubation period the symptom onset event is much more likely to be observed and reported than the infection event).

Reviewer 2: > **What happens if we had the symptom onset date for an individual but not the infection date? Discard their data? Then it's a non-representative sample?

Yes exactly (so no the methods here are not accounting for this) - `epinowcast` has a module for this (the missingness module) but its a tricky problem)

Author: Worth mentioning in this vignette that we don't deal with it then? Or leave that to something more advanced? I'd suggest leave to more advanced.

Reviewer 2: I think very briefly pointing it out (and at some later date linking to something more complex) might be a good user journey

Author: I've tried something brief about this. Hard to know where exactly to put it.

```

With our censored, truncated, and sampled data, we are now ready to attempt to recover the underlying delay distribution using `epidist`.

# Fit the model {#fit}

If we had access to `obs`, then it would be simple to estimate the delay distribution.
However, with only access to `obs_cens_trunc_samp`, we must use a statistical model to avoid severely biased estimates!

The main model function in `epidist` is `latent_truncation_censoring_adjusted_delay`.
<!-- In a future vignette, we will explain in more detail the structure of the model. -->
The parameters of this model are inferred using Bayesian inference.
In particular, we use the the No-U-Turn Sampler (NUTS) Markov chain Monte Carlo (MCMC) algorithm via the `brms` R package.

```{r}
fit <- latent_truncation_censoring_adjusted_delay(
  data = obs_cens_trunc_samp, cores = 4, refresh = 0
)

summary(fit)
```

# Compare estimates {#compare}

Compare modelled estimate to underlying truth here to convince user that it works.
Perhaps as an exercise, we could show the "really simple" estimate being wrong here otherwise it won't be so impressive that the model is right.

> This section should also show the user how to get objects that they might want out of the fitted object.

Reviewer: I think this could be nice, but raises questions about whether we should keep the "naive" model or just code it using brms in the example.

We should also choose a pretty short example delay interval if we want the adjusted estimate to be much better than the raw one.

Author: I think we should code it using `brms` in the example. Perhaps you could say it's the wrong place, but if someone is familiar with `y ~ x` synatx it might reinforce to them that our model is really just some fancy version of that.

Good point about choosing the simulation settings to make sure there is a difference. When I get to this I'll choose the simulation settings to demonstrate the value of using `epidist`.

Reviewer 2: > We should also choose a pretty short example delay interval

Or long distribution?

> I think we should code it using brms in the example.

I think I agree but we may look at this and change our mind if it seems a bit complicated

Reviewer 2: See above comment about maybe doing this via visualing the true dist and the truncated and censored dist. I think do that and then see if we need this? We might still as we can clearly show the summary parameters of the fits vs true (like the old readme but simpler).

Author: I think I've done this.

---
title: "Getting started with epidist"
description: "A quick start guide to using the epidist R package"
output: 

>   bookdown::html_document2:

Reviewer 2: Can you explain?

Author: I find that the `html_vignette2` format creates figures which are formatted strangely (low resolution, incorrect scale of different objects in the figure). I guess this is somewhat of a hack and there is a reason that we would want the vignette to be in `html_vignette2` format rather than `html_document2` format. If there is some way to make the figures less bad in `html_vignette2` please let me know.

Reviewer 2: Okay fair enough. I don't have a strong preference here. I'm not clear if there is a reason for `html_vignette` apart from backwards compatible formatting and if its just that no issue not using

    fig_caption: yes
    code_folding: show
    number_sections: true
pkgdown:
  as_is: true

> # csl: https://raw.githubusercontent.com/citation-style-language/styles/master/apa-numeric-superscript-brackets.csl

Reviewer 2: ?

Author: This was so that "as described in Park et al. ([2024](http://127.0.0.1:36043/rmd_output/0/#ref-park2024estimating)) and Charniga et al. ([2024](http://127.0.0.1:36043/rmd_output/0/#ref-charniga2024best))." appears rather than e.g. [1], [2].

Ideally perhaps the `.bibtex` format would also include linking to the paper. I don't know if any of the `citation-style-language/` ones allow you to do that?

Reviewer 2: Got you. Happy to drop the csl then

    number_sections: true
pkgdown:
  as_is: true
csl: https://raw.githubusercontent.com/citation-style-language/styles/master/apa-numeric-superscript-brackets.csl
# csl: https://raw.githubusercontent.com/citation-style-language/styles/master/apa-numeric-superscript-brackets.csl
link-citations: true
vignette: >
  %\VignetteIndexEntry{Getting Started with Epidist}
  %\VignetteIndexEntry{Getting started with epidist}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: references.bib
---

```{r setup, include=FALSE}
# exclude compile warnings from cmdstanr
knitr::opts_chunk$set(
  fig.path = "figures/getting-started-nowcasting-",
  cache = TRUE, dpi = 330,
  collapse = TRUE, comment = "#>", out.width = "100%",
  message = FALSE, warning = FALSE, error = FALSE,
  eval = FALSE
  fig.path = "figures/epidist-",
  cache = TRUE,
  collapse = TRUE,
  comment = "#>",
  message = FALSE,
  warning = FALSE,
  error = FALSE
)
```

# Quick start

> In this quick start, we demonstrate using `epidist` to ...

Reviewer 2: I quite like a summary - did you drop because you will do later or you don't think its a good idea?

# Package

> In this quick start, we also use `data.table`, `purrr`, and `ggplot2` packages. These are both installed as dependencies when `epidist` is installed. Note that all output from `epidist` is readily useable with  other tools, including `tidyverse` packages (see [here](https://mgimond.github.io/rug_2019_12/Index.html) for a comparison).

Reviewer 2: Why cut?

Author: I don't think this is crucial information for someone reading the "get started" to be presented with at the beginning. If someone cares about dependencies then likely they know that they can look at the `DESCRIPTION` file to figure out the dependencies.

The point about output being readily usable with other tools seems somewhat obvious to me. Are people writing packages where the output is only usable with `data.table` say?

Reviewer 2: > If someone cares about dependencies then likely they know that they can look at the DESCRIPTION file to figure out the dependencies.

Ah so this is because these aren't always going to be package dependencies. We might (and do here - looking at you `gt`) use packages that only appear in the suggests. This means that when people follow our install guides they won't always be able to run the vignette without either installing the suggests or manually knowing about them. 

> The point about output being readily usable with other tools seems somewhat obvious to me. Are people writing packages where the output is only usable with data.table say?

I think it might be obvious to a lot of people but I also think this could scare quite a large minority of people so at the cost of the single sentence seems worth keeping.

Author: Added:

Finally, to run this vignette yourself, you will need the `data.table`, `purrr` and `ggplot2` packages installed.
Note that to work with outputs from `epidist` you do not need to use `data.table`: any tool of your preference is suitable.

<!-- Please see as of yet unwritten vignette for information about how to transform your data of other formats to the right format. -->

First, we use the [Gillepsie algorithm](https://en.wikipedia.org/wiki/Gillespie_algorithm) to generate infectious disease outbreak data (Figure \@ref(fig:outbreak)) from a stochastic compartmental model:

```{r}
outbreak <- simulate_gillespie(seed = 101)
```

(ref:outbreak) Early on in the epidemic, there is a high rate of growth in new cases. As more people are infected, the rate of growth slows. (Only every 50th case is shown to avoid over-plotting.)

```{r outbreak, fig.cap="(ref:outbreak)"}
outbreak[case %% 50 == 0, ] |>
  ggplot(aes(x = ptime, y = case)) +
  geom_point(col = "#56B4E9") +
  labs(x = "Primary event time (day)", y = "Case number") +
  theme_minimal()
```

`outbreak` is a [`data.table`](https://cran.r-project.org/web/packages/data.table/index.html) with the columns `case` and `ptime`.

Now, to generate secondary events, we will use a lognormal distribution (Figure \@ref(fig:lognormal)) for the delay between primary and secondary events:

```{r}
secondary_dist <- data.table(meanlog = 1.8, sdlog = 0.5) |>

>   add_natural_scale_mean_sd()

Reviewer 2: Note there may be some work spinning up on making a small package to carry these transforms around

```

(ref:lognormal) The lognormal distribution is skewed to the right. Long delay times still have some probability.

```{r lognormal, fig.cap="(ref:lognormal)"}
ggplot(data.frame(x = c(0, 8)), aes(x = x)) +
  geom_function(
    fun = dlnorm,
    meanlog = secondary_dist[["meanlog"]],
    sdlog = secondary_dist[["sdlog"]]
  ) +
  theme_minimal() +
  labs(
    x = "Delay between primary and secondary event (days)",
    y = "Probability density"
  )
```

```{r}
obs <- outbreak |>
  simulate_secondary(
    meanlog = secondary_dist[["meanlog"]],
    sdlog = secondary_dist[["sdlog"]]
  )
```

(ref:delay) Secondary events (in green) occur with a delay drawn from the lognormal distribution (Figure \@ref(fig:lognormal)). As with Figure \@ref(fig:outbreak), to make this figure easier to read, only every 50th case is shown.

```{r delay, fig.cap="(ref:delay)"}
obs[case %% 50 == 0, ] |>
  ggplot(aes(y = case)) +
  geom_segment(
    aes(x = ptime, xend = stime, y = case, yend = case),
    col = "grey"
  ) +

>   geom_point(aes(x = ptime), col = "#56B4E9") +

Reviewer 2: I would usually do this by going to long form data and then using the new variable as then you can have a legend. This is nice and clear as is though

Author: I would also usually do that but in some sense I was trying to make the number of lines used low.

  geom_point(aes(x = stime), col = "#009E73") +
  labs(x = "Event time (day)", y = "Case number") +
  theme_minimal()
```

`obs` is now a [`data.table`](https://cran.r-project.org/web/packages/data.table/index.html) object with further columns for `delay` and `stime`.
The secondary event time is simply the primary event time plus the delay:

```{r}
all(obs$ptime + obs$delay == obs$stime)
```

If we were to receive the data `obs` as above then estimating the delay distribution would be easy, and the `epidist` package wouldn't need to exist.
However, in reality, during an outbreak we almost never receive the data as above.

First, the times of primary and secondary events will usually be censored.
This means that rather than exact event times, we observe event times within an interval.
Here we suppose that the interval is daily, meaning that only the date of the primary or secondary event, not the exact event time, is reported (Figure \@ref(fig:cens)):

```{r}
# observe_process() should be renamed and include a "daily" i.e. 1 argument
obs_cens <- obs |> observe_process()
```

> (ref:cens) Interval censoring of the primary and secondary event times obscures the delay times.

Reviewer 2: ```suggestion
(ref:cens) Interval censoring of the primary and secondary event times obscures the delay times. Note that while daily censoring is most common, both the primary and secondary events can have other delay intervals, and this scenario is supported by  `epidist`.
```

Author: Added.

```{r cens, fig.cap="(ref:cens)"}
ggplot(obs_cens, aes(x = delay, y = delay_daily)) +
  geom_point(col = "#E69F00") +
  geom_abline(slope = 1, intercept = 0, linetype = "dashed", col = "grey") +
  theme_minimal() +
  coord_fixed() +
  labs(x = "Exact delay time (days)", y = "Censored delay time (days)")
```

Next, during an outbreak we will usually be estimating delays in real time.
The result is that only those cases with a secondary event occurring before some time will be observed.
This is called (right) truncation, and biases the observation process towards shorter delays:

> <!-- I think possible to do a better job with showing what right truncation is with a figure here! -->

Reviewer 2: yes good idea. We could reproduce one from one of the papers or role a new one. It feels like we could promote this to a issue vs doing here if we think it needs more thought.

In terms of what is in `epidist` I quite like the primary event bar plot with the dots for secondary events but can also see the clear case for it to be done on individual delays when explaining

Author: Going to make an issue to address this.

```{r}
obs_time <- 25
# filter_obs_by_obs_time() should be renamed to refer to stime
obs_cens_trunc <- filter_obs_by_obs_time(obs_cens, obs_time = obs_time)
```

Finally, in reality, it's not possible to observe every case.

> We suppose that a sample of individuals of size `sample_size` are observed:

Reviewer 2: if we wanted to make a point about ascertainment here seems like it could be a good choice

Author: Happy to leave to you and Katie to decide on this. I'd imagine putting ascertainment into a more complex vignette could be better (especially if there is any analysis the package can do to help you with it.).

Reviewer 2: I think we should make a concise point here and link out to something more complex inn the future

```{r}
sample_size <- 200
```

This sample size corresponds to `r 100 * round(sample_size / nrow(obs_cens_trunc), 3)`% of the data.

```{r}
obs_cens_trunc_samp <-
  obs_cens_trunc[sample(seq_len(.N), sample_size, replace = FALSE)]
```

With our censored, truncated, and sampled data, we are now ready to try to recover the underlying delay distribution using `epidist`.

# Fit the model {#fit}

If we had access to the data `obs`, then it would be simple to estimate the delay distribution (Figure \@ref(fig:obs-est)).

> (ref:obs-est) The histogram of delays matches closely the underlying distribution. Edit: at the moment something is wrong here.

Reviewer 2: ruh roh. On the face of it I can't see a code issue. Can dig in if helpful

Author: I'm also confused about why this is wrong.

Author: Yep!

```{r obs-est, fig.cap="(ref:obs-est)"}
ggplot() +
  geom_histogram(data = obs, aes(x = delay, y = stat(count / sum(count)))) +
  geom_function(
    data = data.frame(x = c(0, 30)),
    aes(x = x),
    fun = dlnorm,
    meanlog = secondary_dist[["meanlog"]],
    sdlog = secondary_dist[["sdlog"]],
  ) +
  labs(x = "Delay between primary and secondary event (days)", y = "Probability density") +
  theme_minimal()
```

> However, with only access to `obs_cens_trunc_samp`, we must use a statistical model to avoid severely biased estimates!

Reviewer 2: Prove it to me with a plot

Author: Agree and done.

Author: Still IDK what's going on with the true density..

![image](https://github.com/epinowcast/epidist/assets/29818116/276db2f2-7d4a-48ed-824f-231833c08fe1)

Reviewer 2: Hmm so the version we have here: https://github.com/parksw3/epidist-paper

definitely works as intended (we should talk a bit about why we are plotting from scratch here (is this just for now or is this a long term choice you have made).

Reviewer 2: ah you aren't passing via an args list: https://github.com/parksw3/epidist-paper/blob/d34a461c4d7526438d509eec99025abda30687c7/R/plot.R#L68

do the parameters you are passing influence what is being plot?

Author: Plotting from scratch because it's easier for me to write, rather than it being a long term thought out choice or something like that.

And yes that fixed it, great spot, thanks!

The main `epidist` model function is `latent_truncation_censoring_adjusted_delay`.
<!-- In a future vignette, we will explain in more detail the structure of the model. -->
The parameters of the model are inferred using Bayesian inference.
In particular, we use the the No-U-Turn Sampler (NUTS) Markov chain Monte Carlo (MCMC) algorithm via the [`brms`](https://paul-buerkner.github.io/brms/) R package [@brms].

```{r}
fit <- latent_truncation_censoring_adjusted_delay(
  data = obs_cens_trunc_samp, cores = 4, refresh = 0
)
```

The `fit` object is a `brmsfit` object containing MCMC samples from each of the parameters (Table \@ref(tab:pars)) in the model.
Many of these are so-called "latent variables".

```{r pars}
pars <- fit$fit@par_dims |>

>   map(.f = function(x) if(identical(x, integer(0))) return(1) else return(x))

Reviewer 2: There are other choices that provide more functionality but are heavier like `tidybayes` (heavier in the sense of them themselves having more dependencies).

>   map(.f = function(x) if(identical(x, integer(0))) return(1) else return(x))

Author: I just didn't know an easier way to do this. Open to moving to `brms` / `posterior` / `tidybayes`. This is probably a one off thing for the vignette rather than a function in the package so not sure how much it matters.

>   map(.f = function(x) if(identical(x, integer(0))) return(1) else return(x))

Author: We already have `posterior` in `Imports:`

>   map(.f = function(x) if(identical(x, integer(0))) return(1) else return(x))

Reviewer 2: I think using poosterior might be a good idea. What you have right now is a bit scary

>   map(.f = function(x) if(identical(x, integer(0))) return(1) else return(x))

Author: Will try doing. I don't know the function in `posterior` to do this off the top of my head or if there is one.

>   map(.f = function(x) if(identical(x, integer(0))) return(1) else return(x))

Author: I didn't find an easy way to do this with `posterior`.

data.frame("Parameter" = names(pars), "Length" = unlist(pars)) |>

>   gt::gt(caption = "All of the parameters that are included in the model")

Reviewer 2: As we don't have this as a dep I think we would need to add it as a suggest which I think is a valid thing to do

Author: Yeah I think we will use `gt` again in other vignettes so I'll add it.

```

> Although some users may wish to work with `fit` directly, we will see in the following section that `epidist` provides functions to make common post-processing tasks easy.

Reviewer 2: ```suggestion
Although some users may wish to work with fit directly, we will see in the following section that epidist provides functions to make common post-processing tasks easy. Additionally, users can leverage tools that support brms models for fit objects.
```

My suggestion seems a bit clunky but I more generally think we need to flag this info

Reviewer 2: Also I see from this seciton why you went with manual extraction in the previous section. I like it from a teaching sense my only concern would be does it add more verbosity (answer don't know what do you think?)

Author: Tried:

In the following section, we show that `epidist` provides functions to make common post-processing tasks easy. Additionally, for users familiar with Stan and `brms`, they may like to work with `fit` directly. For example, any tool that supports `brms` fitted model objects will be compatible with `fit`.

# Compare estimates {#compare}

Compare modelled estimate to underlying truth here to convince user that it works.

> Perhaps as an exercise, we could show the "really simple" estimate being wrong here otherwise it won't be so impressive that the model is right.

Reviewer 2: we could either show the simple model is wrong or we could show the truncated dist above more clearly vs the underlying dist? That might do a lot of what fitting a simple model would do?

Author: Yeah having done the above I agree, and "fitting a model" to the obscured data isn't really doing much. It does do some things like teach people we are not doing anything really magic, just an altered `brms` model, but unclear if it's worth it.

Reviewer 2: yeeah agree - also for our impact factor we need to be magical ;)
